File: easy/0121_best_time_to_buy_and_sell_stocks/solution.py
from typing import List

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        min_price = float('inf')
        max_profit = 0

        for price in prices:
            if min_price > price:
                min_price = price

            if max_profit < price - min_price:
                max_profit = price - min_price

        return max_profit

File: easy/0121_best_time_to_buy_and_sell_stocks/solution.cpp
#include <vector>

class Solution {
public:
  int maxProfit(std::vector<int> &prices) {
    const size_t n = prices.size();

    std::vector<int> postfixMaxPrice(n);
    postfixMaxPrice[n - 1] = 0;

    size_t i = n - 1;
    while (i--)
      postfixMaxPrice[i] = std::max(postfixMaxPrice[i + 1], prices[i + 1]);

    int maxProfit = 0;

    for (i = 0; i < n - 1; ++i)
      maxProfit = std::max(maxProfit, postfixMaxPrice[i] - prices[i]);

    return maxProfit;
  }
};
